[SPARK-16240][ML] Model loading backward compatibility for LDA

GayathriMurali opened the pull request:
## What changes were proposed in this pull request?

LDA model loading backward compatibility
## How was this patch tested?

Existing UT

Commit:
[SPARK-16240][ML] Model loading backward compatibility for LDA

GayathriMurali commented:
@hhbyyh Can you please help review? I am not sure if this is the right way to do it, as topicDistributionCol is not included in the MLWriter or load.

SparkQA commented:
**[Test build #62012 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62012/consoleFull)** for PR 14112 at commit [`880c3a1`](https://github.com/apache/spark/commit/880c3a1bfa67101835a5965c65f9f8942e95be35).

SparkQA commented:
**[Test build #62012 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62012/consoleFull)** for PR 14112 at commit [`880c3a1`](https://github.com/apache/spark/commit/880c3a1bfa67101835a5965c65f9f8942e95be35).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62012/
Test FAILed.

GayathriMurali commented:
retest this

hhbyyh commented:
Thanks @GayathriMurali for the PR. I think we'll need to override the default behavior of getAndSetParams. Meanwhile, we need to invoke both convertVectorColumnsToML and convertMatrixColumnsToML. 

I'll send a PR to your repository for reference.

hhbyyh commented:
PR created. Sorry I got something else that I need to turn to. Ideally, the overriding `getAndSetParams` should be in LDAParams, thus it can be reused by LDA and LDA Local/Distributed Model. Please help move it there (perhaps a new Function in LDAParams)

Let me know if you have any question. I'll revisit ASAP.

hhbyyh commented:
@jkbradley I find it not easy to add a unit test to cover the logic. Appreciate your thoughts.

yanboliang commented:
@hhbyyh I think offline test should be OK for now, since we don't have unified save/load compatibility test framework until now. It's better we can get this feature in the next RC.

Commit:
Fixing topicDistributionCol issue

GayathriMurali commented:
@hhbyyh Thanks for helping out. Updated commit includes logic to include topicDistributionCol @yanboliang

SparkQA commented:
**[Test build #62108 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62108/consoleFull)** for PR 14112 at commit [`2b13262`](https://github.com/apache/spark/commit/2b1326258cf3daf436ec54103d34b6b9c0b39be5).

SparkQA commented:
**[Test build #62108 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62108/consoleFull)** for PR 14112 at commit [`2b13262`](https://github.com/apache/spark/commit/2b1326258cf3daf436ec54103d34b6b9c0b39be5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62108/
Test PASSed.

jkbradley commented:
I agree it will be hard to do a proper unit test, but we should definitely test offline.  I've made a JIRA for designing backwards compatibility for the long term: https://issues.apache.org/jira/browse/SPARK-15573

I'll take a look at this now

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -574,18 +578,28 @@ object LocalLDAModel extends MLReadable[LocalLDAModel] {
-      val gammaShape = data.getAs[Double](4)
+      val vectorConverted = MLUtils.convertVectorColumnsToML(data, "docConcentration")
+      val Row(vocabSize: Int, topicsMatrix: Matrix, docConcentration: Vector,
+      topicConcentration: Double, gammaShape: Double) = MLUtils.convertMatrixColumnsToML(
```
style: indent 4 spaces

You could also simplify this with Datasets by using ```(DataFrame).as[Data]``` if you make ```LocalLDAModelWriter.Data``` accessible here (by moving it to sit under ```object LocalLDAModel```).

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -574,18 +578,28 @@ object LocalLDAModel extends MLReadable[LocalLDAModel] {
+           metadata.params match {
+              case JObject(pairs) =>
+                  pairs.foreach { case (paramName, jsonValue) =>
+                      val variable =
```
rename for clarity: "variable" -> "origParam"

jkbradley commented:
What will happen in the future if the model changes, such as storing another val?  We will need to update the model loading logic and should follow different code paths for models saved in 1.6 vs. 2.x.  I'd recommend we isolate the 1.6 code path by checking ```metadata.sparkVersion```.  Even though a single code paths works for 1.6 and 2.0, isolating the special logic for 1.6 might make it easier to update in the future.  What do you think?

jkbradley commented:
Thank for the PR!

hhbyyh commented:
I agree that it will make thing easier to separate the loading logic here in LDA due to its extra complexity. Yet maybe we should not extend the pattern to previous changes about loading compatibility (LR, NaiveBayes, feature). I'd appreciate your suggestions.

@GayathriMurali I think we should cover the DistributedLDAModel and ideally, LDA for loading compatibility and avoid code duplication.

Commit:
Review comments and Seperate coding logic

GayathriMurali commented:
+1 for separate loading logic. The recent commit includes separate code paths depending on sparkVersion

SparkQA commented:
**[Test build #62125 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62125/consoleFull)** for PR 14112 at commit [`08e5b55`](https://github.com/apache/spark/commit/08e5b5566028d794918c94cc375382827e6fd05f).

SparkQA commented:
**[Test build #62125 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62125/consoleFull)** for PR 14112 at commit [`08e5b55`](https://github.com/apache/spark/commit/08e5b5566028d794918c94cc375382827e6fd05f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62125/
Test FAILed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -80,7 +84,8 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
    *     - default = uniformly (1.0 / k), following the implementation from
    *       [[https://github.com/Blei-Lab/onlineldavb]].
-   * @group param
+    *
```
fix indentation

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -121,7 +126,8 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
    *     - default = (1.0 / k), following the implementation from
    *       [[https://github.com/Blei-Lab/onlineldavb]].
-   * @group param
+    *
```
ditto

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -566,26 +565,52 @@ object LocalLDAModel extends MLReadable[LocalLDAModel] {
   }
 
+  private case class Data(
+                           vocabSize: Int,
```
ditto

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -566,26 +565,52 @@ object LocalLDAModel extends MLReadable[LocalLDAModel] {
-      val topicConcentration = data.getAs[Double](3)
-      val gammaShape = data.getAs[Double](4)
+      val vectorConverted = MLUtils.convertVectorColumnsToML(data, "docConcentration")
+      val Row(vocabSize: Int, topicsMatrix: Matrix, docConcentration: Vector,
```
Does this work?  You should need to take the first row

GayathriMurali replied:
It worked when I locally ran the unit tests, but fails here on Jenkins.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -566,26 +565,52 @@ object LocalLDAModel extends MLReadable[LocalLDAModel] {
+             throw new IllegalArgumentException(
+               s"Cannot recognize JSON metadata: ${metadata.metadataJson}.")
+           }
+        case "2.x" =>
```
I don't think you can match on "2.x"  You'll need a more robust comparison of versions.  (Check out what sparkSession.version returns; it could be things like "2.0.1" or "2.0.1-SNAPSHOT".)

jkbradley commented:
@hhbyyh I agree about not changing the other models' loading code unless it becomes necessary.  I hope we can design a better long-term solution during 2.1.

We should definitely cover the other LDA classes.

jkbradley commented:
@GayathriMurali I'd like to accelerate getting this merged.  Please let me know if you'd like help with it, especially with adding the fix for other LDA classes.  (We can collaborate on the same PR.)

GayathriMurali commented:
@jkbradley I am sorry, I have been held up with something else. I am looking on ways to add this to DistribtedLDA model. I will have something by EOD today.

Commit:
Added model loading logic to Distributed LDA and fixed review comments

SparkQA commented:
**[Test build #62293 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62293/consoleFull)** for PR 14112 at commit [`0c2e51c`](https://github.com/apache/spark/commit/0c2e51c2d38207003c5cf659423e71fd2739d003).

SparkQA commented:
**[Test build #62293 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62293/consoleFull)** for PR 14112 at commit [`0c2e51c`](https://github.com/apache/spark/commit/0c2e51c2d38207003c5cf659423e71fd2739d003).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62293/
Test FAILed.

Commit:
Bug fix

SparkQA commented:
**[Test build #62295 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62295/consoleFull)** for PR 14112 at commit [`216777f`](https://github.com/apache/spark/commit/216777fdac275c8865d54e7193aff7e02714cba9).

SparkQA commented:
**[Test build #62295 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62295/consoleFull)** for PR 14112 at commit [`216777f`](https://github.com/apache/spark/commit/216777fdac275c8865d54e7193aff7e02714cba9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62295/
Test FAILed.

Commit:
Bug fix

SparkQA commented:
**[Test build #62298 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62298/consoleFull)** for PR 14112 at commit [`58384d4`](https://github.com/apache/spark/commit/58384d447de9f1fd5959c9bfe0caae2e4bac92ae).

SparkQA commented:
**[Test build #62298 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62298/consoleFull)** for PR 14112 at commit [`58384d4`](https://github.com/apache/spark/commit/58384d447de9f1fd5959c9bfe0caae2e4bac92ae).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62298/
Test FAILed.

Commit:
Selecting columns directly instead of as[Data]- Unit test issue

SparkQA commented:
**[Test build #62337 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62337/consoleFull)** for PR 14112 at commit [`8848f0e`](https://github.com/apache/spark/commit/8848f0e9b4fa57b6ce819d0b5bc47e94a8a0bb2e).

SparkQA commented:
**[Test build #62337 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62337/consoleFull)** for PR 14112 at commit [`8848f0e`](https://github.com/apache/spark/commit/8848f0e9b4fa57b6ce819d0b5bc47e94a8a0bb2e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62337/
Test PASSed.

GayathriMurali commented:
@jkbradley I implemented model loading logic for DistributedLDA as well. I am using a versionRegex  for robustness in version checking. Using `as[Data].head()` is producing a scala match error on Jenkins but not when I run locally on my machine . For now I removed that and I am using `.select(....).head()`  and all test cases are passing on Jenkins. I am investigating why it should fail. Meanwhile, it would be great if you can leave your comments on the recent commits.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -566,26 +565,54 @@ object LocalLDAModel extends MLReadable[LocalLDAModel] {
+                          topicsMatrix: Matrix,
+                          docConcentration: Vector,
+                          topicConcentration: Double,
+                          gammaShape: Double)
```
indent

jkbradley replied:
Ping: fix indentation

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -728,16 +755,40 @@ object DistributedLDAModel extends MLReadable[DistributedLDAModel] {
+          }
+        case pattern =>
+          DefaultParamsReader.getAndSetParams(model, metadata)
+      }
```
We'd better find a way to reuse the code.

GayathriMurali replied:
@hhbyyh Thank you! I can maybe write this as a separate function.

Commit:
Added loading logic to LDA and wrapped topicdistribution loading in a separate function

SparkQA commented:
**[Test build #62567 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62567/consoleFull)** for PR 14112 at commit [`a8bdd7a`](https://github.com/apache/spark/commit/a8bdd7a748e63a1948f98130a704972b9cbca54b).

SparkQA commented:
**[Test build #62567 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62567/consoleFull)** for PR 14112 at commit [`a8bdd7a`](https://github.com/apache/spark/commit/a8bdd7a748e63a1948f98130a704972b9cbca54b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62567/
Test FAILed.

Commit:
Style issues

SparkQA commented:
**[Test build #62568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62568/consoleFull)** for PR 14112 at commit [`b16b368`](https://github.com/apache/spark/commit/b16b3682a50486a81913ce5b5b67711d73b7589f).

SparkQA commented:
**[Test build #62568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62568/consoleFull)** for PR 14112 at commit [`b16b368`](https://github.com/apache/spark/commit/b16b3682a50486a81913ce5b5b67711d73b7589f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62568/
Test PASSed.

GayathriMurali commented:
@jkbradley Can you please help review this?

GayathriMurali commented:
@jkbradley Please let me know if I can do anything to help get this merged

GayathriMurali commented:
@jkbradley Can you please help review this?

SparkQA commented:
**[Test build #3234 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3234/consoleFull)** for PR 14112 at commit [`b16b368`](https://github.com/apache/spark/commit/b16b3682a50486a81913ce5b5b67711d73b7589f).

jkbradley commented:
@GayathriMurali Apologies for the long delay!  It slipped past the release, and I'm trying to catch up on PRs now.  I'll make a final review pass ASAP

SparkQA commented:
**[Test build #3234 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3234/consoleFull)** for PR 14112 at commit [`b16b368`](https://github.com/apache/spark/commit/b16b3682a50486a81913ce5b5b67711d73b7589f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -26,29 +29,27 @@ import org.apache.spark.ml.linalg.{Matrix, Vector, Vectors, VectorUDT}
 import org.apache.spark.mllib.impl.PeriodicCheckpointer
-import org.apache.spark.mllib.linalg.{Matrices => OldMatrices, Vector => OldVector,
-  Vectors => OldVectors}
+import org.apache.spark.mllib.linalg.{Matrices => OldMatrices, Vector => OldVector, Vectors => OldVectors}
```
OldMatrices is not used

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -26,29 +29,27 @@ import org.apache.spark.ml.linalg.{Matrix, Vector, Vectors, VectorUDT}
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.{DataFrame, Dataset, Row, SparkSession}
-import org.apache.spark.sql.functions.{col, monotonically_increasing_id, udf}
+import org.apache.spark.sql.functions.{col, monotonicallyIncreasingId, udf}
```
revert this; monotonically_increasing_id is preferred as of 2.0

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -26,29 +29,27 @@ import org.apache.spark.ml.linalg.{Matrix, Vector, Vectors, VectorUDT}
-   *
-   * @group param
-   */
+   /*
```
Fix indentation, and use proper Java doc comments starting with ```/**```

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -57,31 +58,32 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
-   *       [[https://github.com/Blei-Lab/onlineldavb]].
-   * @group param
-   */
+   /*
```
ditto: fix

It looks like autoformatting has messed up doc strings elsewhere below as well.  Can you please check through and fix those?

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -352,6 +355,25 @@ private[clustering] trait LDAParams extends Params with HasFeaturesCol with HasM
   }
+
+  def extractTopicDistributionCol(metadata: Metadata, model: Params): Unit =
+  {
```
style: brace should go on previous line.  Please compare with existing code when unsure about style

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -566,26 +581,41 @@ object LocalLDAModel extends MLReadable[LocalLDAModel] {
+      metadata.sparkVersion match {
+        case "1.6" =>
+          model.extractTopicDistributionCol(metadata, model)
+        case versionRegex =>
```
This is not doing regex pattern matching.  It is creating a temp val "versionRegex" storing the value of metadata.sparkVersion; it matches on everything.

I would recommend using a regex to extract the ```\d+\.\d+``` major and minor version pattern and then doing a match-case on that extracted value.  You could create a helper method for extracting the version so that you do not duplicate code.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/clustering/LDA.scala:
```diff
@@ -728,16 +758,27 @@ object DistributedLDAModel extends MLReadable[DistributedLDAModel] {
+      metadata.sparkVersion match {
+        case "1.6" =>
+          model.extractTopicDistributionCol(metadata, model)
+        case pattern =>
```
This too is not doing regex pattern matching.

jkbradley commented:
Done with review for now; I'll check back for updates.  Thanks!

jkbradley commented:
@GayathriMurali Btw, I know it's been a long time.  If you no longer have time to work on this, please say so, and I'd be happy to take it over.  Thanks!

GayathriMurali commented:
@jkbradley  I am so sorry I couldn't respond to this on time! I am in a transition process and might not be able to drive this JIRA to completion at this point in time. Thanks!

jkbradley commented:
@GayathriMurali No problem; you've had to wait quite a while.  I'll be happy to take it over.

Could you please close this PR for now?

I'll send a new PR based on your commits (so you'll still be the primary author).  Thanks!

GayathriMurali commented:
@jkbradley Sure! Thanks.
